src/Middleware/MessageRouterMiddleware.php

```diff
@@ -0,0 +1,118 @@
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * This software consists of voluntary contributions made by many individuals
+ * and is licensed under the MIT license.
+ */
+
+namespace ZfrEbWorker\Middleware;
+
+use Psr\Container\ContainerInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Webimpress\HttpMiddlewareCompatibility\HandlerInterface as DelegateInterface;
+use Webimpress\HttpMiddlewareCompatibility\MiddlewareInterface;
+use ZfrEbWorker\Exception\InvalidArgumentException;
+use ZfrEbWorker\Exception\RuntimeException;
+
+/**
+ * MessageRouter middleware
+ * What this thing does is Routing to a specific handler based on message name.
+ * You can find a complete reference of what Elastic Beanstalk set here:
+ * http://docs.aws.amazon.com/elasticbeanstalk/latest/dg/using-features-managing-env-tiers.html
+ *
+ * @author Michaël Gallego
+ */
+class MessageRouterMiddleware implements MiddlewareInterface
+{
+    const MESSAGE_NAME_ATTRIBUTE = 'worker.message_name';
+
+    /**
+     * @var ContainerInterface
+     */
+    private $container;
+
+    /**
+     * Map message names to a middleware name. For instance:
+     * [
+     *     'image.saved' => ProcessImageMiddleware::class
+     * ]
+     *
+     * @var array
+     */
+    private $messagesMapping;
+
+    /**
+     * @param array              $messagesMapping
+     * @param ContainerInterface $container
+     */
+    public function __construct(array $messagesMapping, ContainerInterface $container)
+    {
+        $this->messagesMapping = $messagesMapping;
+        $this->container       = $container;
+    }
+
+    /**
+     * @param ServerRequestInterface $request
+     * @param DelegateInterface      $delegate
+     *
+     * @return ResponseInterface
+     */
+    public function process(ServerRequestInterface $request, DelegateInterface $delegate): ResponseInterface
+    {
+        $name = $request->getAttribute(WorkerMiddleware::MESSAGE_NAME_ATTRIBUTE);
+
+        // Let's create a middleware pipeline of mapped middlewares
+        $middleware = $this->getMiddlewareForMessage($name);
+        $response   = $middleware->process($request, $delegate);
+
+        // Some middleware may return a 204 or any other 2xx answer, which are considered as success. However Elastic Beanstalk is picky
+        // and only accept 200 OK to delete message. So we normalize any status in the 2xx range
+        $statusCode = $response->getStatusCode();
+
+        if ($statusCode >= 200 && $statusCode <= 299) {
+            $response = $response->withStatus(200);
```

Comment:
The status code normalization and `X-Handled-By` header could stay in `WorkerMiddleware` since they are unrelated to routing.